ibis/expr/timecontext.py

```diff
@@ -0,0 +1,191 @@
+""" Time context module
+
+This is an attempt to allow us to implement the time context extension without
+affecting the existing SQL-like execution model for backends.
+
+Most of the execution is built on the foundation that "Data is uniquely
+defined by the op tree". This is true in SQL analysis where there is no
+ambiguity what the result of executing a TableExpr is.
+
+In time series analysis, however, this is not necessarily True. We have defined
+an extension to ibis execution for time series analysis where the result of
+executing a TableExpr is defined by the TableExpr plus the time context are
+associated with the execution.
+
+Time context specifies the temporal range of a query, it carries the start and
+end datetimes. For example, a TableExpr can represent the query select count(a)
+from table, but the result of that is different with time context
+("20190101", "20200101") vs ("20200101", "20210101"), because what data is in
+"table" depends also on the time context.
+
+While data in scope is public and global for all nodes, timecontext is intended
+to store 'local' time context data for each node in execution. i.e., each
+subtree of an expr tree can have different time context. Which makes it so
+that when executing each node, we also need to know the "local time context"
+for that node.
+
+And we propose to store these data as 'timecontext', calculate in execution
+pass it along to children nodes, in the ibis tree. See each backends for
+implementation details.
+"""
+
+import enum
+import functools
+from typing import Optional
+
+import pandas as pd
+
+import ibis.common.exceptions as com
+import ibis.expr.api as ir
+import ibis.expr.operations as ops
+from ibis.expr.typing import TimeContext
+
+# In order to use time context feature, there must be a column of Timestamp
+# type, and named as 'time' in TableExpr. This TIME_COL constant will be
+# used in filtering data from a table or columns of a table.
+TIME_COL = 'time'
+
+
+class TimeContextRelation(enum.Enum):
+    """ Enum to classify the relationship between two time contexts
+        Assume that we have two timecontext c1 (begin1, end1),
```

Comment:
Indentation seems off here. Docstring doesn't need indentation from the second line.

Reply:
+1 Fixed it.